### PR TITLE
Horizon tests

### PR DESCRIPTION
**Pull request by HollowPrincess**

### CODE REVIEW

File: seismiqb/tests/geometry_test.py [hunk 1]
```diff
@@ -1,19 +1,50 @@
-""" Script for running notebook with SeismicGeometry tests."""
+""" Script for running notebook with SeismicGeometry tests.
```
[the remark this answers is not in the document]
Comment (HollowPrincess) [reply seismiqb/tests/geometry_test.py:1]: It runs only the controller notebook, which runs others. Both formulations are correct. But the main functionality of this script is to run the **controller notebook**.

As an alternative, I propose ```Script for running the controller notebook for SeismicGeometry tests.```

File: seismiqb/tests/geometry_test.py [hunk 2]
```diff
@@ -1,19 +1,50 @@
...
+
+The test contains some global constants:
+
+DATESTAMP : str
+    Execution date in "YYYY-MM-DD" format.
+    It is used for saving notebooks executions and temporary files.
+TESTS_SCRIPTS_DIR : str
+    Path to the directory with tests .py scripts.
+    It is used for providing paths to files for execution.
```
[the remark this answers is not in the document]
Comment (HollowPrincess) [reply seismiqb/tests/geometry_test.py:10]: It is rightly noted.
Does 
```
TESTS_SCRIPTS_DIR : str
    Path to the directory with test .py scripts.
    Used as an entry point to the working directory.
```
sound better?

File: seismiqb/tests/horizon_test.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        if SHOW_TEST_ERROR_INFO:
+            # Add error traceback into the message
+            msg.append(extract_traceback(path_ipynb=out_path_ipynb))
+
+    with capsys.disabled():
+        # Tests output
+        if SHOW_MESSAGE:
+            for line in msg:
+                print(line)
+        else:
```
[the remark this answers is not in the document]
Comment (HollowPrincess) [reply seismiqb/tests/horizon_test.py:143]: I want to get the last line in the message for getting the info about tests' success or failure. The last line is used in line number 153.

File: seismiqb/src/labels/horizon.py [hunk 4]
```diff
@@ -939,8 +939,10 @@ def evaluate(self, compute_metric=True, supports=50, plot=True, savepath=None, p
     def compare(self, other, offset=0, absolute=True, printer=print, hist=True, plot=True):
         """ Compare quality of self against another horizon or sequence of horizons. """
         from ..metrics import HorizonMetrics
-        HorizonMetrics([self, other]).evaluate('compare', absolute=absolute, offset=offset,
-                                               printer=printer, hist=hist, plot=plot)
+        HorizonMetrics([self, other]).evaluate(
+            'compare', absolute=absolute, offset=offset, printer=printer,
+            hist=hist, plot=plot
```
Comment (SergeyTsimfer) [inline seismiqb/src/labels/horizon.py:944]: Why?

File: seismiqb/src/labels/horizon_attributes.py [hunk 5]
```diff
@@ -226,7 +225,7 @@ def filled_matrix(self):
 
     def grad_along_axis(self, axis=0):
         """ Change of heights along specified direction. """
-        grad = np.diff(self.matrix, axis=axis, prepend=np.int32(0))
+        grad = np.diff(self.matrix, axis=axis, prepend=self.FILL_VALUE)
```
Comment (SergeyTsimfer) [inline seismiqb/src/labels/horizon_attributes.py:228]: Can you explain the reason behind it? Better yet, make some images to explain the difference and where it fixes the behaviour

File: seismiqb/tests/geometry_test.py [hunk 6]
```diff
@@ -1,19 +1,52 @@
...
+
+DATESTAMP : str
+    Execution date in "YYYY-MM-DD" format.
+    Used for saving notebooks executions and temporary files.
+TESTS_SCRIPTS_DIR : str
+    Path to the directory with test .py scripts.
+    Used as an entry point to the working directory.
+NOTEBOOKS_DIR : str
+    Path to the directory with test .ipynb files.
+TEST_DIR : str
```
Comment (SergeyTsimfer) [inline seismiqb/tests/geometry_test.py:13]: `OUTPUT_DIR`, maybe? using the word `test` inside files with tests is usually bad practice (here and everywhere else) 
:)

File: seismiqb/tests/horizon_test.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+def test_horizon(capsys, tmpdir):
+    """ Run Horizon test notebook.
+
+    This test runs ./notebooks/horizon_test.ipynb test file and show execution message.
+
+    Under the hood, this notebook create a fake seismic cube with horizon, saves them
+    and runs Horizon tests notebooks (base, manipulations, attributes).
+    """
+    # Delete old test notebook results
+    if GITHUB_MODE:
```
Comment (SergeyTsimfer) [inline seismiqb/tests/horizon_test.py:84]: What do you think about always using the `tmpdir`, but copying the files to an `OUTPUT_DIR` if not `REMOVE_EXTRA_FILES` afterward? that would reduce the amount of code and possibly make it cleaner

**Review by SergeyTsimfer — COMMENTED** [review]
Overall, looks great
🎉

**Review by cdtn — COMMENTED** [review]
1. Use `pylint_notebook`.
2. Check horizon derived attributes reset when points/matrix storage updated.

File: seismiqb/src/labels/horizon_attributes.py (hunk 5 above)
Comment (SergeyTsimfer) [inline seismiqb/src/labels/horizon_attributes.py:228]: Is that still needed, considering changes in #143?

**Review by SergeyTsimfer — APPROVED** [review]

**Pull request merged**